test/unit/lib/utils/fs/safeMoveFile.test.js

```diff
@@ -0,0 +1,160 @@
+'use strict';
+
+const fse = require('fs-extra');
+const sinon = require('sinon');
+const chai = require('chai');
+const { getTmpDirPath } = require('../../../../utils/fs');
+const { join } = require('path');
+const { expect } = require('chai');
+const fsp = require('fs').promises;
+const sinonChai = require('sinon-chai');
+const safeMoveFile = require('../../../../../lib/utils/fs/safeMoveFile');
+
+chai.use(sinonChai);
+
+/**
+ * Returns and Error object that resembles the EXDEV errors that are thrown
+ * when you try to rename across device boundaries.
+ *
+ * @returns an Error object that resembles the EXDEV errors are thrown
+ */
+const createFakeExDevError = () => {
+  // Properties captured by using fs.renameSync in the Node v12.20.1 REPL
+  const fakeCrossDeviceError = new Error(
+    "Error: EXDEV: cross-device link not permitted, rename '/foo/bar' -> '/bar/baz'"
+  );
+  // This is the important property that safeMoveFile looks for
+  // to fallback to copy then rename behaviour
+  fakeCrossDeviceError.code = 'EXDEV';
+
+  // These other properties aren't used but it doesn't hurt to have an accurate error object
+  fakeCrossDeviceError.errno = -18;
+  fakeCrossDeviceError.syscall = 'rename';
+  fakeCrossDeviceError.path = '/foo/bar';
+  fakeCrossDeviceError.dest = '/bar/baz';
+  return fakeCrossDeviceError;
+};
+
+/**
+ * The name of the file that will be moved
+ */
+const artefactName = 'foo-bar.zip';
+
+// Generate temporary directories once per test run
+// They will be cleaned up when the tests are done
+
+/**
+ * The folder that the test file will be moved to
+ */
+const destinationPath = getTmpDirPath();
+
+/**
+ * The folder that the test file will come from
+ */
+const sourcePath = getTmpDirPath();
+
+/**
+ * The full path to destination where the file is being moved to
+ */
+const destinationFile = join(destinationPath, artefactName);
+
+/**
+ * The full path to the file that is being moved
+ */
+const sourceFile = join(sourcePath, artefactName);
+
+describe('#safeMoveFile()', () => {
+  let renameStub;
+  let copyFileStub;
+  let unlinkStub;
+
+  beforeEach(async () => {
+    renameStub = sinon.stub(fsp, 'rename');
+    copyFileStub = sinon.stub(fsp, 'copyFile');
+    unlinkStub = sinon.stub(fsp, 'unlink');
+    // Allow the fs calls to act as normal until we want to force the rename to fail
+    renameStub.callThrough();
+    copyFileStub.callThrough();
+    unlinkStub.callThrough();
```

Comment:
Ideally if in test we treat implementation as _black box_ and with above approach, we're technically confirming that implementation is calling given functions and how many times. We practised that in some parts and it appeared as very problematic for refactors (where we improve internals without affecting the outcome).

I propose, to change a bit methodology:
- Let's mock just `rename` (as without that, it's hard to (1) confirm rename is used for instant files move (2) emualate `'EXDEV'` error), and let all others `fs` functions work naturally
- In each test simply confirm that file at source path no longer exists, expected file exists at target path, `rename` was called to move it.

Reply:
That makes a lot of sense. I've updated the test to what you proposed.